### ISSUE TITLE
Update pysaml2 in catalog

### ISSUE THREAD

**Issue by thejuliekramer** [GSA/data.gov#1416]
## Feature

As a part of a [plan to upgrade CKAN extensions](https://docs.google.com/spreadsheets/d/1LgVvgiUC5Vc-OQaSmtvQNlTFLhs3Mho65ekggYEiYh4/edit?ts=5e50447f#gid=0) in use, and in order to improve the way that we manage forks we need to update the use of [pysaml2](https://github.com/GSA/pysaml2) in the catalog.

We need to replace the use of outdated pysaml2 wherever it occurs with a newer version.

## Tasks
- [x] Determine if anything is missing in the upstream that we want to bring over from the GSA version 
- [ ] Update pysaml2 and any configuration
- [ ] Update datagov-deploy with any updated configuration
- [ ] Do a lot of testing. :)

## Acceptance Criteria
- [ ] PR replacing usage of pysaml2 repo
- [ ] [catalog](https://github.com/GSA/catalog-app) builds correctly and functionality remains the same

**Comment by thejuliekramer** [comment]
@adborden @mogul @avdata99 let me know what you think about this.

List of steps here:

- [ ] [remove the importing of pysaml2 in the inventory branch](https://github.com/GSA/ckan/pull/73) 
- [ ] import it where it is used in the [ckanext-saml2 repository](https://github.com/GSA/ckanext-saml2). In that repository we should import the [upstream version](https://github.com/IdentityPython/pysaml2), because the [GSA fork](https://github.com/GSA/pysaml2) is only [two commits ahead ](https://github.com/IdentityPython/pysaml2/compare/master...tobes:master)- and the commits were made by someone outside of the data.gov team and seem suspect. Instead of using the github version we can use the [PiPy library](https://pypi.org/project/pysaml2/) and import it as a package - this is being done by @avdata99 in a separate PR, I will link to it when I receive it.
- [ ] remove the import  of pysaml2 from the [inventory-app](https://github.com/GSA/inventory-app/pull/27) once the [ckanext-saml2 repository](https://github.com/GSA/ckanext-saml2) is updated with the import

**Comment by adborden** [comment]
@thejuliekramer that sounds right, although at the moment, the ckan fork doesn't need to be updated, we're using requirements.txt in inventory-app and the one in ckan is ignored.

For the two commits, it looks like @FuhuXia accepted those. @FuhuXia do you know why [this fix](https://github.com/GSA/pysaml2/commit/fd0396d81cfe6abd42b8fe50d677f4d3f34485b3) was needed?

**Comment by avdata99** [comment]
This library is used by saml2 extension. The plan in #1403 is to stop using this fork and move to upstream python library. This issue depends on #1403 definition and tests.  

We need to include opinion/analysis of @thejuliekramer in this fork changes. Is there something in this fork that we really need to use? I think no, just to be sure.

If the changes in this fork are not important and saml2 ext works fine without them we will be able to close this issue.

**Comment by starsinmypockets** [comment]
I don't think we can load the upstream dependency directly due to max-specific configuration in code: 
https://github.com/IdentityPython/pysaml2/compare/master...GSA:max

The upstream hasn't gone very far since it was forked by (tobes and then) GSA: 
https://github.com/IdentityPython/pysaml2/compare/master...GSA:master

The required changes to rebase look fine:  https://github.com/IdentityPython/pysaml2/compare/master...GSA:master

I think the same approach that we took with `ckanext-saml2` is in order: rebase onto the IdentityPython version, listen for changes upstream, rinse and repeat

**Comment by starsinmypockets** [comment]
@adborden can you quickly review my previous comment and I'll push a new branch `datagov/5.0.0` to `gsa/pysaml2` ?

**Comment by adborden** [comment]
Quoting Paul Walker (2020-03-30 16:34:03)
> @adborden can you quickly review my previous comment and I'll push a new branch `datagov/5.0.0` to `gsa/pysaml2` ?

Let's sync up on this tomorrow, Andres and Julie seemed to think the
changes to the pysaml2 fork were not necessary. Can you document what
the changes are and why they would/wouldn't be necessary?

--
Aaron D Borden
Human and hacker

**Comment by starsinmypockets** [comment]
This looks to me like the relevant diff between the latest upstream and max:

https://github.com/IdentityPython/pysaml2/compare/master...GSA:max

The changes in `max` are related, as the branch-name suggests, to the non-standard MAX.gov interface, and I don't believe that we can safely lose them.

This would suggest using the same approach we used for `ckanext-saml2` -- rebase changes onto the latest upstream branch, and watch for releases upstream (rinse and repeat)
 
Still need to determine if the latest upstream supports python 2

**Comment by starsinmypockets** [comment]
According to pypi 4.9 is the last version with python 2 support
https://pypi.org/project/pysaml2/4.9.0/

**Comment by starsinmypockets** [comment]
We can still watch `v4.x.y` for fixes
https://github.com/IdentityPython/pysaml2/compare/v4.9.0...GSA:max

**Comment by starsinmypockets** [comment]
( @adborden please review )

The [relevant diff here](https://github.com/GSA/pysaml2/compare/GSA:max...IdentityPython:v4.9.0), however, is massive -- ~2000 commits - rebase will not be practical in this case. The directory structure has changed & etc) so I think we can follow 2 courses of action:

Option 1. Fork and reapply our commit manually, then deploy and manually QA in staging

Option 2. (Still evaluating the practicality of this approach) Add an adaptor module -- a separate repo / plugin that overrides the relevant functions needed for MAX.gov and can be loaded after `pysaml2@v4.9.0` in our apps -- the advantage here is that we don't maintain a fork of the pysaml2 lib, and we can add automated tests to test that the monkey-patched functions return the expected result... failure will turn CI red and get the attention of devs, otherwise, smooth sailing. 

In either case we should watch for 4.x.y releases (in case of hotfix without new tag), and 4.x.y releases and rerun CI (Or automate this if practical)

**Comment by starsinmypockets** [comment]
I do prefer option 2, above, as it allows us to make testable assertions which will catch relevant changes to the upstream, and captures the spirit of the hack for future devs

Basically the failure path looks like:
1. CI fails with test failure
2. Dev investigates and sees that signature verification test fails (eg)
3. Dev looks at relevant changes in upstream and updates our adaptor or takes appropriate action

**Comment by adborden** [comment]
@starsinmypockets you have the compare reversed, it's [only 10 commits](https://github.com/IdentityPython/pysaml2/compare/v4.9.0...GSA:max).

**Comment by adborden** [comment]
Why do you want to work off of v4.9.0 and not [v5.0.0](https://github.com/IdentityPython/pysaml2/releases/tag/v5.0.0)?

**Comment by starsinmypockets** [comment]
@adborden The doesn't make sense to me given the intervening time, as well as the different directory structure. Maybe I'm missing something about the way this is calculated. Either way, I'm honestly not sure how to bring the changes forward to the latest commit upstream using git alone.

v4.9.0 appears to be the last branch that supports python 2.7

**Comment by starsinmypockets** [comment]
@adborden Here's a stubbed version of an adaptor that we can test:

This should allow us to stop maintaing an additional fork (granted that we maintain this small plugin...)
https://github.com/datopian/max_pysaml2

**Comment by adborden** [comment]
Ah, I didn't see a reference to python3 only in the 5.0.0 docs. Okay, let's pair on this today.

**Comment by estebanruseler** [comment]
Update document to be discussed with Aaron and Tom & co https://hackmd.io/TuQLq6WrSRaS9gmDWyzU1g

**Comment by adborden** [comment]
@kimwdavidson you were right, the description of this work isn't accurate. I updated the description to remove references to Inventory and to PyPI. Instead, we're rebasing our fork to a more recent version of pysaml2 (still not the latest but as far as we can go with ~CKAN 2.3~ python 2.7) and only updating for catalog.
